File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/metrics/audio/SpeakerAudioType.java
package com.mapbox.services.android.navigation.v5.navigation.metrics.audio;

import android.content.Context;
import android.media.AudioManager;

public class SpeakerAudioType implements AudioTypeResolver {
  private static final String SPEAKER = "speaker";
  private AudioTypeResolver chain;

  @Override
  public void nextChain(AudioTypeResolver chain) {
    this.chain = chain;
  }

  @Override
  public String obtainAudioType(Context context) {
    AudioManager audioManager = (AudioManager) context.getSystemService(Context.AUDIO_SERVICE);
    if (audioManager == null) {
      return "unknown";
    }
    return audioManager.isSpeakerphoneOn() ? SPEAKER : chain.obtainAudioType(context);
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/metrics/audio/SpeakerAudioType.kt
package com.mapbox.services.android.navigation.v5.navigation.metrics.audio

import android.content.Context
import android.media.AudioManager

internal class SpeakerAudioType : AudioTypeResolver {
    private val SPEAKER = "speaker"
    private var chain: AudioTypeResolver? = null

    override fun nextChain(chain: AudioTypeResolver) {
        this.chain = chain
    }

    override fun obtainAudioType(context: Context): String {
        val audioManager:AudioManager? =  context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
        return audioManager?.let {audio->
            when (audio.isSpeakerphoneOn){
                true -> SPEAKER
                false ->{
                    chain?.obtainAudioType(context) ?:SPEAKER
                }
            }
        }?:SPEAKER
    }
}